File: apps/android/app/src/main/java/bot/molt/android/networking/ApiModels.kt
package bot.molt.android.networking

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

// MARK: - Agent Models

@Serializable
data class Agent(
    val id: String,
    val name: String,
    val status: AgentStatus,
    val agentId: String? = null,
    val createdAt: String,
    val updatedAt: String,
    val organizationId: String,
    val containers: List<AgentContainer>? = null,
    val channels: List<AgentChannel>? = null,
    val healthChecks: List<AgentHealthCheck>? = null,
)

@Serializable
enum class AgentStatus {
    PENDING, PROVISIONING, RUNNING, STOPPED, ERROR, DELETING, TERMINATED
}

@Serializable
data class AgentContainer(
    val id: String,
    val status: String,
    val cpu: Double? = null,
    val memory: Int? = null,
    val gatewayUrl: String? = null,
)

@Serializable
data class AgentChannel(
    val id: String,
    @SerialName("channelType") val type: String,
    val name: String? = null,
    @SerialName("isActive") val enabled: Boolean,
)

@Serializable
data class AgentHealthCheck(
    val id: String,
    val isHealthy: Boolean,
    val responseTime: Int? = null,
    val checkedAt: String? = null,
)

// MARK: - Organization Models

@Serializable
data class Organization(
    val id: String,
    val name: String,
    val slug: String? = null,
    val logo: String? = null,
    val isPersonal: Boolean? = null,
)

@Serializable
data class OrganizationMember(
    val id: String,
    val userId: String,
    val role: String,
    val user: MemberUser? = null,
)

@Serializable
data class MemberUser(
    val id: String,
    val name: String? = null,
    val email: String,
    val image: String? = null,
)

// MARK: - Task Models

@Serializable
data class AgentTask(
    val id: String,
    val title: String,
    val description: String? = null,
    val status: TaskStatus,
    val labels: List<String>? = null,
    val taskType: String? = null,
    val urgency: Int? = null,
    val agentId: String,
    val deliverables: String? = null,
    val deliveredAt: String? = null,
    val delegationStatus: String? = null,
    val delegatedToAgentId: String? = null,
    val createdAt: String,
    val updatedAt: String,
)

@Serializable
enum class TaskStatus {
    todo, in_progress, delivered, done, archived
}

// MARK: - Crew Models

@Serializable
data class Crew(
    val id: String,
    val organizationId: String,
    val name: String,
    val description: String? = null,
    val status: CrewStatus,
    val leadAgentId: String,
    val leadAgent: CrewAgentRef? = null,
    val members: List<CrewMember>? = null,
    val createdAt: String,
    val updatedAt: String,
)

@Serializable
enum class CrewStatus {
    active, paused, disbanded
}

@Serializable
data class CrewMember(
    val id: String,
    val agentId: String,
    val role: String,
    val agent: CrewAgentRef? = null,
)

@Serializable
data class CrewAgentRef(
    val id: String,
    val name: String,
    val status: AgentStatus,
)

@Serializable
data class CrewExecution(
    val id: String,
    val crewId: String,
    val status: String,
    val objective: String? = null,
    val startedAt: String? = null,
    val completedAt: String? = null,
    val createdAt: String,
)

// MARK: - Attachment Models

@Serializable
data class TaskAttachment(
    val id: String,
    val taskId: String,
    val filename: String,
    val mimeType: String,
    val fileSize: Int,
    val source: String? = null,
    val createdAt: String,
)

// MARK: - Chat Models

@Serializable
data class ChatMessage(
    val id: String,
    val content: String,
    val role: String,
    val agentId: String,
    val userId: String? = null,
    val createdAt: String,
)

// MARK: - RPC Input/Output

@Serializable
data class ListAgentsInput(val organizationId: String)

@Serializable
data class AgentIdInput(val id: String, val organizationId: String)

@Serializable
data class CreateAgentInput(
    val organizationId: String,
    val name: String,
    val personaId: String? = null,
    val podId: String? = null,
)

@Serializable
data class AgentActionInput(val agentId: String, val organizationId: String)

@Serializable
data class AgentListResponse(val agents: List<Agent>)

@Serializable
data class AgentResponse(
    val agent: Agent,
    val hasLlmCredentials: Boolean? = null,
    val configuredCredentials: List<String>? = null,
)

@Serializable
data class OrganizationListResponse(val organizations: List<Organization>)

@Serializable
data class ChatHistoryInput(val agentId: String, val organizationId: String, val limit: Int? = null)

@Serializable
data class ChatHistoryResponse(
    val messages: List<ChatMessage>,
    val total: Int? = null,
    val hasMore: Boolean? = null,
)

@Serializable
data class ChatSaveInput(val agentId: String, val organizationId: String, val role: String, val content: String)

@Serializable
data class TaskListResponse(val tasks: List<AgentTask>)

@Serializable
data class TaskListInput(val agentId: String, val organizationId: String)

@Serializable
data class LogEntry(val id: String, val level: String, val message: String, val timestamp: String)

@Serializable
data class LogsInput(val agentId: String, val organizationId: String)

@Serializable
data class LogsResponse(val logs: List<LogEntry>)

// MARK: - Crew RPC Types

@Serializable
data class ListCrewsInput(val organizationId: String)

@Serializable
data class CrewIdInput(val crewId: String, val organizationId: String)

@Serializable
data class CrewListResponse(val crews: List<Crew>)

@Serializable
data class CrewResponse(val crew: Crew)

@Serializable
data class CrewExecutionListResponse(val executions: List<CrewExecution>)

@Serializable
data class CrewExecutionTasksInput(val executionId: String, val organizationId: String)

@Serializable
data class StartCrewExecutionInput(val crewId: String, val organizationId: String, val objective: String? = null)

// MARK: - Composio Integration Models

@Serializable
data class ComposioApp(
    val id: String,
    val name: String,
    val displayName: String? = null,
    val logo: String? = null,
    val categories: List<String>? = null,
)

@Serializable
data class ComposioConnection(
    val id: String,
    val appName: String,
    val status: String,
)

@Serializable
data class ComposioAppsInput(val organizationId: String, val oauthOnly: Boolean? = null)

@Serializable
data class ComposioAppsResponse(val apps: List<ComposioApp>, val total: Int? = null)

@Serializable
data class ComposioConnectionsInput(val organizationId: String)

@Serializable
data class ComposioConnectionsResponse(val connections: List<ComposioConnection>)

@Serializable
data class ComposioConnectInput(val organizationId: String, val appName: String)

@Serializable
data class ComposioConnectResponse(val redirectUrl: String? = null, val connectionId: String? = null, val connected: Boolean? = null)

@Serializable
data class ComposioDisconnectInput(val organizationId: String, val appName: String)

// MARK: - Attachment RPC Types

@Serializable
data class AttachmentListInput(val agentId: String, val organizationId: String)

@Serializable
data class TaskAttachmentListInput(val taskId: String, val organizationId: String)

@Serializable
data class AttachmentListResponse(val attachments: List<TaskAttachment>)

@Serializable
data class AttachmentDownloadInput(val attachmentId: String, val taskId: String, val organizationId: String)

@Serializable
data class AttachmentDownloadResponse(val downloadUrl: String, val filename: String, val mimeType: String)

File: apps/ios/Sources/Networking/APIModels.swift
import Foundation

// MARK: - Agent Models

struct Agent: Codable, Identifiable, Sendable {
    let id: String
    let name: String
    let status: AgentStatus
    let agentId: String?
    let createdAt: String
    let updatedAt: String
    let organizationId: String
    let containers: [AgentContainer]?
    let channels: [AgentChannel]?
    let healthChecks: [AgentHealthCheck]?
}

enum AgentStatus: String, Codable, Sendable {
    case PENDING
    case PROVISIONING
    case RUNNING
    case STOPPED
    case ERROR
    case DELETING
    case TERMINATED // legacy/fallback
}

struct AgentContainer: Codable, Identifiable, Sendable {
    let id: String
    let status: String
    let cpu: Double?
    let memory: Int?
    let gatewayUrl: String?
}

struct AgentChannel: Codable, Identifiable, Sendable {
    let id: String
    let type: String
    let name: String?
    let enabled: Bool

    private enum CodingKeys: String, CodingKey {
        case id
        case type = "channelType"
        case name
        case enabled = "isActive"
    }
}

struct AgentHealthCheck: Codable, Sendable {
    let id: String
    let isHealthy: Bool
    let responseTime: Int?
    let checkedAt: String?
}

// MARK: - Organization Models

struct Organization: Codable, Identifiable, Sendable {
    let id: String
    let name: String
    let slug: String?
    let logo: String?
    let isPersonal: Bool?
}

struct OrganizationMember: Codable, Identifiable, Sendable {
    let id: String
    let userId: String
    let role: String
    let user: MemberUser?
}

struct MemberUser: Codable, Sendable {
    let id: String
    let name: String?
    let email: String
    let image: String?
}

// MARK: - Task Models

struct AgentTask: Codable, Identifiable, Sendable {
    let id: String
    let title: String
    let description: String?
    let status: TaskStatus
    let labels: [String]?
    let taskType: String?
    let urgency: Int?
    let agentId: String
    let deliverables: String?
    let deliveredAt: String?
    let delegationStatus: String?
    let delegatedToAgentId: String?
    let createdAt: String
    let updatedAt: String
}

enum TaskStatus: String, Codable, Sendable {
    case todo
    case in_progress
    case delivered
    case done
    case archived
}

// MARK: - Crew Models

struct Crew: Codable, Identifiable, Sendable {
    let id: String
    let organizationId: String
    let name: String
    let description: String?
    let status: CrewStatus
    let leadAgentId: String
    let leadAgent: CrewAgentRef?
    let members: [CrewMember]?
    let createdAt: String
    let updatedAt: String
}

enum CrewStatus: String, Codable, Sendable {
    case active
    case paused
    case disbanded
}

struct CrewMember: Codable, Identifiable, Sendable {
    let id: String
    let agentId: String
    let role: String
    let agent: CrewAgentRef?
}

struct CrewAgentRef: Codable, Sendable {
    let id: String
    let name: String
    let status: AgentStatus
}

struct CrewExecution: Codable, Identifiable, Sendable {
    let id: String
    let crewId: String
    let status: String
    let objective: String?
    let startedAt: String?
    let completedAt: String?
    let createdAt: String
}

// MARK: - Attachment Models

struct TaskAttachment: Codable, Identifiable, Sendable {
    let id: String
    let taskId: String
    let filename: String
    let mimeType: String
    let fileSize: Int
    let source: String?
    let createdAt: String
}

// MARK: - Credential Models

struct CredentialInfo: Codable, Identifiable, Sendable {
    let id: String
    let name: String
    let provider: String
    let hasValue: Bool
}

// MARK: - Chat Models

struct ChatMessage: Codable, Identifiable, Sendable {
    let id: String
    var content: String
    let role: String
    let agentId: String
    let userId: String?
    let createdAt: String
    var isStreaming: Bool?

    private enum CodingKeys: String, CodingKey {
        case id, content, role, agentId, userId, createdAt
    }
}

// MARK: - Chat Event Models (Socket.IO streaming)

struct ChatEventPayload: Sendable {
    let sessionKey: String?
    let state: String // "delta", "final", "error"
    let role: String?
    let text: String?

    /// Parse from raw Socket.IO `agent:event` dictionary
    init?(from dict: [String: Any]) {
        guard let eventType = dict["event"] as? String, eventType == "chat",
              let payload = dict["payload"] as? [String: Any]
        else { return nil }

        self.sessionKey = payload["sessionKey"] as? String
        self.state = payload["state"] as? String ?? "delta"

        // Extract text from message.content[].text
        if let message = payload["message"] as? [String: Any] {
            self.role = message["role"] as? String
            if let contentArray = message["content"] as? [[String: Any]] {
                self.text = contentArray.compactMap { $0["text"] as? String }.joined()
            } else if let contentStr = message["content"] as? String {
                self.text = contentStr
            } else {
                self.text = nil
            }
        } else if let content = payload["content"] as? String {
            self.role = payload["role"] as? String
            self.text = content
        } else {
            self.role = nil
            self.text = nil
        }
    }
}

// MARK: - Chat Attachment Models

struct ChatAttachmentInput: Sendable {
    let mimeType: String
    let content: String // base64
    let fileName: String?

    var asDictionary: [String: String] {
        var dict: [String: String] = ["mimeType": mimeType, "content": content]
        if let fileName { dict["fileName"] = fileName }
        return dict
    }
}

struct ChatSaveInput: Codable {
    let agentId: String
    let organizationId: String
    let role: String
    let content: String
}

struct ChatHistoryInput: Codable {
    let agentId: String
    let organizationId: String
    let limit: Int?
}

struct ChatHistoryFullResponse: Codable {
    let messages: [ChatMessage]
    let total: Int?
    let hasMore: Bool?
}

// MARK: - RPC Input Structs

struct ListAgentsInput: Codable {
    let organizationId: String
}

struct AgentIdInput: Codable {
    let id: String
    let organizationId: String
}

struct CreateAgentInput: Codable {
    let organizationId: String
    let name: String
    let personaId: String?
    let podId: String?
}

struct AgentActionInput: Codable {
    let agentId: String
    let organizationId: String
}

struct CreateTaskInput: Codable {
    let agentId: String
    let organizationId: String
    let title: String
    let description: String?
    let status: String
    let taskType: String
    let urgency: Int
}

struct UpdateTaskInput: Codable {
    let taskId: String
    let organizationId: String
    let title: String?
    let description: String?
    let status: String?
    let urgency: Int?
}

// MARK: - RPC Response Wrappers

struct AgentListResponse: Codable {
    let agents: [Agent]
}

struct AgentResponse: Codable {
    let agent: Agent
    let hasLlmCredentials: Bool?
    let configuredCredentials: [String]?
}

struct TaskListResponse: Codable {
    let tasks: [AgentTask]
}

struct TaskResponse: Codable {
    let task: AgentTask
}

struct OrganizationListResponse: Codable {
    let organizations: [Organization]
}

struct ChatHistoryResponse: Codable {
    let messages: [ChatMessage]
}

// MARK: - Crew RPC Types

struct ListCrewsInput: Codable {
    let organizationId: String
}

struct CrewIdInput: Codable {
    let crewId: String
    let organizationId: String
}

struct CrewListResponse: Codable {
    let crews: [Crew]
}

struct CrewResponse: Codable {
    let crew: Crew
}

struct CrewExecutionListResponse: Codable {
    let executions: [CrewExecution]
}

struct CrewExecutionTasksInput: Codable {
    let executionId: String
    let organizationId: String
}

struct StartCrewExecutionInput: Codable {
    let crewId: String
    let organizationId: String
    let objective: String?
}

// MARK: - Composio Integration Models

struct ComposioApp: Codable, Identifiable, Sendable {
    let id: String
    let name: String
    let displayName: String?
    let logo: String?
    let categories: [String]?
}

struct ComposioConnection: Codable, Identifiable, Sendable {
    let id: String
    let appName: String
    let status: String
}

struct ComposioAppsInput: Codable {
    let organizationId: String
    let oauthOnly: Bool?
}

struct ComposioAppsResponse: Codable {
    let apps: [ComposioApp]
    let total: Int?
}

struct ComposioConnectionsInput: Codable {
    let organizationId: String
}

struct ComposioConnectionsResponse: Codable {
    let connections: [ComposioConnection]
}

struct ComposioConnectInput: Codable {
    let organizationId: String
    let appName: String
}

struct ComposioConnectResponse: Codable {
    let redirectUrl: String?
    let connectionId: String?
    let connected: Bool?
}

struct ComposioDisconnectInput: Codable {
    let organizationId: String
    let appName: String
}

// MARK: - Attachment RPC Types

struct AttachmentListInput: Codable {
    let agentId: String
    let organizationId: String
}

struct TaskAttachmentListInput: Codable {
    let taskId: String
    let organizationId: String
}

struct AttachmentListResponse: Codable {
    let attachments: [TaskAttachment]
}

struct AttachmentDownloadInput: Codable {
    let attachmentId: String
    let taskId: String
    let organizationId: String
}

struct AttachmentDownloadResponse: Codable {
    let downloadUrl: String
    let filename: String
    let mimeType: String
}
